File: test/find_block_test/find_block_new_test.cpp
//#include <stdafx.h>
//#include<algorithm> // sort(), std::max()
//#include<functional> // std::less, std::greater..
//#include<ciso646> // for "and, not, or, ..."
//#include<iostream>
#include<FQlib.h>

void find_block_(const pts_info *infos, const double radius, const double *bound_y, const double *bound_x, int *block_mask)
{
	int i, j, k, lb, lby, lb_d, lb_d_, seq = 0;
	int idy = infos->idy;// block id of the point
	int idx = infos->idx;
	double y = infos->y;// the coordinates of the point
	double x = infos->x;
	double scale = infos->scale;// the length of the side of the square blocks
	int ny = infos->ny; // the number of blocks along each axis 
	int nx = infos->nx;
	int num = infos->blocks_num;// the numbers of total blocks

	int nx_left, nx_right, ny_up, ny_down; // the max blocks number in each direction
												// away from the point within radius_e
	int nx_s, nx_e, ny_s, ny_e;

	double re_sq = radius * radius;
	double dx, dy;
	double distance[4];
	// "distance" stores the distance of the four vertexes of each block,
	// the sequence of the vertexes£º
	// | (y1,x1), (y1,x2) |
	// | (y2,x1), (y2,x2) |

	// find the minimum square that contains the target blocks
	nx_left = (int)((radius - x + bound_x[0]) / scale) + idx + 1;
	nx_right = (int)((radius + x - bound_x[0]) / scale) - idx;
	ny_up = (int)((radius + y - bound_y[0]) / scale) - idy;
	ny_down = (int)((radius - y + bound_y[0]) / scale) + idy + 1;

	nx_s = std::max(idx - nx_left, 0);
	nx_e = std::min(idx + nx_right + 1, nx);
	ny_s = std::max(idy - ny_down, 0);
	ny_e = std::min(idy + ny_up + 1, ny);

	// initialiize the mask
	for (i = 0; i < num; i++)
	{
		block_mask[i] = -1;
	}

	for (i = ny_s; i < ny_e; i++)
	{
		lby = i * nx; // for speed
		for (j = nx_s; j < nx_e; j++)
		{
			lb = lby + j; // label of the block
			lb_d = lb * 4;
			for (k = 0; k < 4; k++)
			{
				lb_d_ = lb_d + k;
				dy = bound_y[lb_d_] - y;
				dx = bound_x[lb_d_] - x;
				distance[k] = dy * dy + dx * dx;
			}
			sort_arr(distance, 4, 1); // ascending order
			if (distance[0] > re_sq and i not_eq idy and j not_eq idx)
			{
				//	"distance[0] > re_sq and i not_eq idy and j not_eq idx":  
				//  if the minimum distance between the vertexes of a block ,
				//	which is not on the cross centers on  (idy, idx) , is larger 
				//	than radius_e, it is not the target one. 
				//	The blocks on the cross centers on  (idy, idx) are the 
				//	targets.
				continue;
			}
			else
			{
				block_mask[seq] = lb;
				seq++;
			}
		}
	}
}

int main(int argc, char* argv[])
{
	int ny, nx;
	int i, j, k;
	int tag, bin_label;
	double scale = 0.15;
	double x, y;
	double rad_s, rad_e;
	int shape[2];
	int area_id = 1, num;
	int ra_bin_num, dec_bin_num;

	char path[50], data_path[200];
	char buf[100], set_name[50], attr_name[50];


	// sky area, the label of the foreground galaxy and the radius for searching
	area_id = std::atoi(argv[1]);
	tag = std::atoi(argv[2]);
	rad_e = std::atof(argv[3]);


	// for the mask of needed grid
	sprintf(data_path, "/mnt/ddnfs/data_users/hkli/CFHT/gg_lensing/data/cata_result_ext_grid.hdf5");
	sprintf(set_name, "/background/w_%d", area_id);
	sprintf(attr_name, "grid_shape");
	read_h5_attrs(data_path, set_name, attr_name, shape, "g");
	ny = shape[0];
	nx = shape[1];
	int *mask = new int[ny*nx]{};
	std::cout << "Grid: " << ny << " " << nx << " " << scale << std::endl;
	// the boundary of blocks
	sprintf(set_name, "/background/w_%d/block_boundx", area_id);
	sprintf(attr_name, "shape");	
	read_h5_attrs(data_path, set_name, attr_name, shape, "d");
	double *boundx = new double[shape[0]*shape[1]]{};
	read_h5(data_path, set_name, boundx);

	sprintf(set_name, "/background/w_%d/block_boundy", area_id);
	sprintf(attr_name, "shape");
	read_h5_attrs(data_path, set_name, attr_name, shape, "d");
	double *boundy = new double[shape[0] * shape[1]]{};
	read_h5(data_path, set_name, boundy);

	sprintf(set_name, "/background/w_%d/RA_bin", area_id);
	sprintf(attr_name, "shape");
	read_h5_attrs(data_path, set_name, attr_name, shape, "d");
	double *ra_bin = new double[shape[0] * shape[1]]{};
	read_h5(data_path, set_name, ra_bin);
	ra_bin_num = shape[0] - 1;

	sprintf(set_name, "/background/w_%d/DEC_bin", area_id);
	sprintf(attr_name, "shape");
	read_h5_attrs(data_path, set_name, attr_name, shape, "d");
	double *dec_bin = new double[shape[0] * shape[1]]{};
	read_h5(data_path, set_name, dec_bin);
	dec_bin_num = shape[0] - 1;

	
	// the foreground 
	sprintf(set_name, "/foreground/w_%d/RA", area_id);
	sprintf(attr_name, "shape");
	read_h5_attrs(data_path, set_name, attr_name, shape, "d");
	double *ra = new double[shape[0]]{};
	read_h5(data_path, set_name, ra);

	sprintf(set_name, "/foreground/w_%d/DEC", area_id);
	sprintf(attr_name, "shape");
	read_h5_attrs(data_path, set_name, attr_name, shape, "d");
	double *dec = new double[shape[0]]{};
	read_h5(data_path, set_name, dec);

	sprintf(set_name, "/foreground/w_%d/COS_DEC", area_id);
	sprintf(attr_name, "shape");
	read_h5_attrs(data_path, set_name, attr_name, shape, "d");
	double *cos_dec = new double[shape[0]]{};
	read_h5(data_path, set_name, cos_dec);

	num = shape[0];

	if (tag >= num)
	{
		tag = num - 1;
	}

	pts_info pts;

	// the information of the foreground galaxy
	histogram2d_s(dec[tag], ra[tag], dec_bin, ra_bin, dec_bin_num, ra_bin_num, bin_label);
	std::cout << dec_bin_num << " " << ra_bin_num <<" "<<bin_label<< std::endl;
	show_arr(dec_bin, 1, dec_bin_num + 1);
	std::cout << std::endl;
	show_arr(ra_bin, 1, ra_bin_num + 1);
	std::cout << std::endl;

	pts.idx = bin_label % nx;
	pts.idy = bin_label / nx;
	pts.nx = nx;
	pts.ny = ny;
	pts.scale = scale;
	pts.blocks_num = ny * nx;
	pts.x = ra[tag];
	pts.y = dec[tag];
	std::cout << "Position: "<<dec[tag]<<" "<<ra[tag] <<" Grid: "<<pts.idy<< " " << pts.idx<< " (" << pts.ny<<" "<<pts.nx <<")"<< std::endl;

	find_block(&pts, rad_e, boundy, boundx, mask);

	std::cout << std::endl;
	sprintf(buf, "PTS: %d, %d", pts.idy, pts.idx);
	std::cout << buf << std::endl;
	std::cout << scale << std::endl;
	std::cout << pts.y << ", " << pts.x << std::endl;


	sprintf(path, "!mask.fits");
	write_fits(path, mask, ny, nx);

	delete[] boundx;
	delete[] boundy;
	delete[] mask;
	delete[] ra;
	delete[] dec;
	delete[] cos_dec;
	gsl_free();
	return 0;
}

File: test/find_block_test/find_block_new_test.py
import numpy
from sys import path, argv
import os
my_home = os.popen("echo $MYWORK_DIR").readlines()[0][:-1]
path.append('%s/work/mylib/'%my_home)
import plot_tool
import matplotlib.pyplot as plt
from astropy.io import fits
import h5py


area_id = int(argv[1])
tag = int(argv[2])
radius_e = float(argv[3])

scale = 0.15

h5f = h5py.File("/mnt/ddnfs/data_users/hkli/CFHT/gg_lensing/data/cata_result_ext_grid.hdf5","r")
shape = h5f["/background/w_%d"%area_id].attrs["grid_shape"]
ny, nx = shape[0], shape[1]
ra = h5f["/foreground/w_%d/RA"%area_id].value[tag]
dec = h5f["/foreground/w_%d/DEC"%area_id].value[tag]

ra_bin = h5f["/background/w_%d/RA_bin"%area_id].value
dec_bin = h5f["/background/w_%d/DEC_bin"%area_id].value

ra_min, ra_max = ra_bin.min(), ra_bin.max()
dec_min, dec_max = dec_bin.min(), dec_bin.max()

h5f.close()
print(ra,dec,ra_bin[1] - ra_bin[0]-scale, dec_bin[1] - dec_bin[0]-scale)


mask = fits.open("mask.fits")[0].data
target_blocks = []
for i in range(ny):
    for j in range(nx):
        if mask[i,j] > -1:
            m,n = divmod(mask[i,j], nx)
            target_blocks.append((m,n))
            print(m,n,dec_bin[m], dec_bin[m+1], ra_bin[n], ra_bin[n+1])
print(len(target_blocks))
img = plot_tool.Image_Plot(fig_x=20, fig_y=int(20.*ny/nx))
img.plot_img(1,1)

for i in range(ny + 1):
    img.axs[0][0].plot([ra_min, ra_max], [dec_bin[i], dec_bin[i]], c="black", linewidth=0.5)
    for j in range(nx + 1):
        img.axs[0][0].plot([ra_bin[j], ra_bin[j]], [dec_min, dec_max], c="black", linewidth=0.5)

for blks in target_blocks:
    blk_x, blks_y = ra_bin[blks[1]] + 0.5 * scale, dec_bin[blks[0]] + 0.5 * scale
    img.axs[0][0].scatter(blk_x, blks_y, s=5, c="blue")
img.axs[0][0].scatter(ra, dec, s=10, c="red")

theta = numpy.linspace(0,2*numpy.pi,10000)
re_y, re_x = dec + radius_e * numpy.sin(theta), ra + radius_e * numpy.cos(theta)
ey1 = re_y >= dec_min
ey2 = re_y <= dec_max
ex1 = re_x >= ra_min
ex2 = re_x <= ra_max
idx2 = ey1 & ey2 & ex1 & ex2

img.axs[0][0].scatter(re_x[idx2], re_y[idx2], c="coral", s=0.03)
limit = None
if limit:
    x_min, x_max = max(ra - 1.5*radius_e,ra_min), min(ra + 1.5*radius_e, ra_max)
    y_min, y_max = max(dec - 1.5*radius_e, dec_min), min(dec + 1.5*radius_e, dec_max)

    img.axs[0][0].set_xlim(x_min, x_max)
    img.axs[0][0].set_ylim(y_min, y_max)
img.save_img("radius.png")
